# <span style="color:#6E2C00"> Lab feature engineering </span>

>I am adding todays cleaning to previous labs cleaning. 

In [1]:
#Import Libraries
import pandas as pd
import numpy as np
import datetime
import warnings
import matplotlib.pyplot as plt
import seaborn as sns

#Settings
warnings.filterwarnings('ignore')
pd.set_option('display.max_columns', None) # display all columns
pd.set_option('display.max_rows', None) #Show all the rows

In [2]:
# 🔧 Basic functions (by: ISI)
def open_data(data): # returns shape, data types & shows a small sample
    print(f"Data shape is {data.shape}.")
    print()
    print(data.dtypes)
    print()
    print("Data row sample and full columns:")
    return data.sample(5)

def snake_columns(data): # snake_case columns
    data.columns = [column.lower().replace(' ', '_') for column in data.columns]
    return data.sample(0)

def explore_data(data): # sum & returns duplicates, NaN & empty spaces
    duplicate_rows = data.duplicated().sum()
    nan_values = data.isna().sum()
    empty_spaces = data.eq(' ').sum()
    import pandas as pd
    exploration = pd.DataFrame({"NaN": nan_values, "EmptySpaces": empty_spaces}).sort_values('NaN', ascending = False) 
    print(f"There are {data.duplicated().sum()} duplicate rows. Also;")
    return exploration

# <span style="color:#7B241C"> 01 Data import </span>

In [3]:
#Import data
data = pd.read_csv('learningSet.csv')
open_data(data)

Data shape is (95412, 481).

ODATEDW       int64
OSOURCE      object
TCODE         int64
STATE        object
ZIP          object
MAILCODE     object
PVASTATE     object
DOB           int64
NOEXCH       object
RECINHSE     object
RECP3        object
RECPGVG      object
RECSWEEP     object
MDMAUD       object
DOMAIN       object
CLUSTER      object
AGE         float64
AGEFLAG      object
HOMEOWNR     object
CHILD03      object
CHILD07      object
CHILD12      object
CHILD18      object
NUMCHLD     float64
INCOME      float64
GENDER       object
WEALTH1     float64
HIT           int64
MBCRAFT     float64
MBGARDEN    float64
MBBOOKS     float64
MBCOLECT    float64
MAGFAML     float64
MAGFEM      float64
MAGMALE     float64
PUBGARDN    float64
PUBCULIN    float64
PUBHLTH     float64
PUBDOITY    float64
PUBNEWFN    float64
PUBPHOTO    float64
PUBOPP      float64
DATASRCE     object
MALEMILI      int64
MALEVET       int64
VIETVETS      int64
WWIIVETS      int64
LOCALGOV      int64
STATEGOV   

ODATEDW OSOURCE  TCODE STATE    ZIP MAILCODE PVASTATE   DOB NOEXCH  \
47579     9401     DBL      1    LA  70119                    2408      0   
12613     8601     BHG      2    CA  93657                    4709      0   
36660     9601     IWD      0    AZ  85719                    4801      0   
4108      8901     CNL      0    TX  77066                    6001      0   
10342     9601     CRG      0    CO  80804                    2601      0   

      RECINHSE RECP3 RECPGVG RECSWEEP MDMAUD DOMAIN CLUSTER   AGE AGEFLAG  \
47579                                   XXXX     U3       9  73.0       E   
12613                                   XXXX     R1      42  50.0       E   
36660                                   XXXX     S3      21  50.0       E   
4108                                    XXXX     C1      24  38.0       E   
10342                                   XXXX     R3      53  72.0       E   

      HOMEOWNR CHILD03 CHILD07 CHILD12 CHILD18  NUMCHLD  INCOME GENDER  \
47579        U                                      NaN     2.0      M   
12613        H                               F      2.0     4.0      M   
36660        H                                      NaN     6.0      M   
4108         H               M                      1.0     7.0      F   
10342        U                                      NaN     2.0      F   

       WEALTH1  HIT  MBCRAFT  MBGARDEN  MBBOOKS  MBCOLECT  MAGFAML  MAGFEM  \
47579      2.0    0      NaN       NaN      NaN       NaN      NaN     NaN   
12613      4.0   24      1.0       1.0      7.0       0.0      2.0     0.0   
36660      NaN    0      NaN       NaN      NaN       NaN      NaN     NaN   
4108       9.0   25      0.0       0.0      3.0       0.0      0.0     0.0   
10342      1.0   11      1.0       0.0      3.0       0.0      0.0     0.0   

       MAGMALE  PUBGARDN  PUBCULIN  PUBHLTH  PUBDOITY  PUBNEWFN  PUBPHOTO  \
47579      NaN       NaN       NaN      NaN       NaN       NaN       NaN   
12613      0.0       2.0       2.0      2.0       3.0       0.0       0.0   
36660      NaN       NaN       NaN      NaN       NaN       NaN       NaN   
4108       0.0       0.0       0.0      0.0       0.0       4.0       0.0   
10342      0.0       0.0       0.0      0.0       1.0       2.0       0.0   

       PUBOPP DATASRCE  MALEMILI  MALEVET  VIETVETS  WWIIVETS  LOCALGOV  \
47579     NaN        1         0        6        33         0         4   
12613     0.0        3         0       27        15        37        10   
36660     NaN        2         0        0         0         0         0   
4108      9.0        3         0       47        19        34         7   
10342     1.0        3         0       35        20        50         5   

       STATEGOV  FEDGOV SOLP3 SOLIH MAJOR  WEALTH2 GEOCODE COLLECT1 VETERANS  \
47579         4       6                        2.0                             
12613         5       4                        5.0                         Y   
36660         0       0                        NaN                             
4108          3       2                        9.0       4                     
10342         8       2                        NaN                             

      BIBLE CATLG HOMEE PETS CDPLAY STEREO PCOWNERS PHOTO CRAFTS FISHER  \
47579                                                                     
12613     Y                Y             Y        Y            Y          
36660                                                                     
4108                                                                      
10342                                                                     

      GARDENIN BOATS WALKER KIDSTUFF CARDS PLATES LIFESRC PEPSTRFL  POP901  \
47579                                                            X     630   
12613        Y            Y                             3        X    1730   
36660                                                                  786   
4108     

# <span style="color:#7B241C">  02 Data cleaning </span>

In [4]:
explore_data(data)

There are 0 duplicate rows. Also;


NaN  EmptySpaces
RDATE_5   95403            0
RAMNT_5   95403            0
RDATE_3   95170            0
RAMNT_3   95170            0
RDATE_4   95131            0
RAMNT_4   95131            0
RAMNT_6   94636            0
RDATE_6   94636            0
RAMNT_15  88150            0
RDATE_15  88150            0
RDATE_23  87553            0
RAMNT_23  87553            0
RDATE_20  87524            0
RAMNT_20  87524            0
RAMNT_7   86517            0
RDATE_7   86517            0
RAMNT_17  86011            0
RDATE_17  86011            0
RDATE_21  85899            0
RAMNT_21  85899            0
RAMNT_10  84951            0
RDATE_10  84951            0
RDATE_13  83162            0
RAMNT_13  83162            0
NUMCHLD   83026            0
RDATE_11  80672            0
RAMNT_11  80672            0
RDATE_19  79535            0
RAMNT_19  79535            0
RAMNT_9   78678            0
RDATE_9   78678            0
RDATE_24  77674            0
RAMNT_24  77674            0
RAMNT_18  75634            0
RDATE_18  75634            0
RAMNT_22  74539            0
RDATE_22  74539            0
RAMNT_8   73940            0
RDATE_8   73940            0
RDATE_14  72095            0
RAMNT_14  72095            0
RAMNT_12  69712            0
RDATE_12  69712            0
RAMNT_16  68418            0
RDATE_16  68418            0
ADATE_15  65477            0
ADATE_23  56270            0
MBCOLECT  52914            0
MAGFAML   52854            0
MBCRAFT   52854            0
MBBOOKS   52854            0
MAGFEM    52854            0
MAGMALE   52854            0
PUBGARDN  52854            0
PUBCULIN  52854            0
PUBHLTH   52854            0
PUBDOITY  52854            0
PUBNEWFN  52854            0
PUBPHOTO  52854            0
PUBOPP    52854            0
MBGARDEN  52854            0
ADATE_20  50200            0
WEALTH1   44732            0
WEALTH2   43823            0
ADATE_13  40219            0
ADATE_24  36973            0
ADATE_21  35212            0
ADATE_5   33590            0
ADATE_10  32748            0
ADATE_17  27650            0
ADATE_22  25648            0
ADATE_19  24480            0
AGE       23665            0
INCOME    21286            0
ADATE_18  21263            0
ADATE_16  20364            0
ADATE_14  18867            0
ADATE_9   11245            0
ADATE_11  10422            0
NEXTDATE   9973            0
TIMELAG    9973            0
ADATE_12   8923            0
ADATE_7    8874            0
ADATE_6    3557            0
ADATE_8    3511            0
ADATE_4    2191            0
ADATE_3    1950            0
MSA         132            0
ADI         132            0
DMA         132            0
GEOCODE2    132          187
CLUSTER2    132            0
NUMPROM       0            0
MAXADATE      0            0
AFC6          0            0
AFC5          0            0
MDMAUD_F      0            0
MDMAUD_A      0            0
AFC4          0            0
AFC3          0            0
AFC2          0            0
AFC1          0            0
SEC5          0            0
SEC4          0            0
SEC3          0            0
SEC2          0            0
SEC1          0            0
EC8           0            0
EC7           0            0
EC6           0            0
VC1           0            0
VC2           0            0
VC3           0            0
ANC9          0            0
POBC1         0            0
ANC15         0            0
ANC14         0            0
ANC13         0            0
ANC12         0            0
ANC11         0            0
ANC10         0            0
ANC8          0            0
VC4           0            0
ANC7          0            0
ANC6          0            0
ANC5          0            0
ANC4          0            0
ANC3          0            0
ANC2          0            0
ANC1          0            0
EC5           0            0
EC4           0            0
EC3           0            0
OCC10         0            0
EIC4          0            0
EIC3          0            0
EIC2          0            0
EIC1      

>Check for null values in the numerical columns.

In [5]:
data_numeric=data.select_dtypes(include=['float','int'])
explore_data(data_numeric)

There are 0 duplicate rows. Also;


NaN  EmptySpaces
RDATE_5   95403            0
RAMNT_5   95403            0
RDATE_3   95170            0
RAMNT_3   95170            0
RDATE_4   95131            0
RAMNT_4   95131            0
RAMNT_6   94636            0
RDATE_6   94636            0
RDATE_15  88150            0
RAMNT_15  88150            0
RDATE_23  87553            0
RAMNT_23  87553            0
RAMNT_20  87524            0
RDATE_20  87524            0
RAMNT_7   86517            0
RDATE_7   86517            0
RDATE_17  86011            0
RAMNT_17  86011            0
RAMNT_21  85899            0
RDATE_21  85899            0
RDATE_10  84951            0
RAMNT_10  84951            0
RDATE_13  83162            0
RAMNT_13  83162            0
NUMCHLD   83026            0
RAMNT_11  80672            0
RDATE_11  80672            0
RDATE_19  79535            0
RAMNT_19  79535            0
RDATE_9   78678            0
RAMNT_9   78678            0
RAMNT_24  77674            0
RDATE_24  77674            0
RAMNT_18  75634            0
RDATE_18  75634            0
RDATE_22  74539            0
RAMNT_22  74539            0
RAMNT_8   73940            0
RDATE_8   73940            0
RAMNT_14  72095            0
RDATE_14  72095            0
RAMNT_12  69712            0
RDATE_12  69712            0
RDATE_16  68418            0
RAMNT_16  68418            0
ADATE_15  65477            0
ADATE_23  56270            0
MBCOLECT  52914            0
PUBHLTH   52854            0
PUBDOITY  52854            0
PUBPHOTO  52854            0
PUBOPP    52854            0
PUBCULIN  52854            0
PUBGARDN  52854            0
MAGMALE   52854            0
MAGFEM    52854            0
MAGFAML   52854            0
MBBOOKS   52854            0
MBGARDEN  52854            0
MBCRAFT   52854            0
PUBNEWFN  52854            0
ADATE_20  50200            0
WEALTH1   44732            0
WEALTH2   43823            0
ADATE_13  40219            0
ADATE_24  36973            0
ADATE_21  35212            0
ADATE_5   33590            0
ADATE_10  32748            0
ADATE_17  27650            0
ADATE_22  25648            0
ADATE_19  24480            0
AGE       23665            0
INCOME    21286            0
ADATE_18  21263            0
ADATE_16  20364            0
ADATE_14  18867            0
ADATE_9   11245            0
ADATE_11  10422            0
NEXTDATE   9973            0
TIMELAG    9973            0
ADATE_12   8923            0
ADATE_7    8874            0
ADATE_6    3557            0
ADATE_8    3511            0
ADATE_4    2191            0
ADATE_3    1950            0
ADI         132            0
MSA         132            0
DMA         132            0
CLUSTER2    132            0
CARDPROM      0            0
EC3           0            0
EC5           0            0
EC6           0            0
EC7           0            0
EC8           0            0
SEC1          0            0
SEC2          0            0
SEC3          0            0
SEC4          0            0
SEC5          0            0
AFC1          0            0
AFC2          0            0
AFC3          0            0
AFC4          0            0
AFC5          0            0
AFC6          0            0
VC1           0            0
VC2           0            0
VC3           0            0
VC4           0            0
ANC1          0            0
ANC2          0            0
ANC3          0            0
ANC4          0            0
ANC5          0            0
ANC6          0            0
ANC7          0            0
ANC8          0            0
EC4           0            0
EC2           0            0
ANC10         0            0
EC1           0            0
OCC10         0            0
OCC11         0            0
OCC12         0            0
OCC13         0            0
EIC1          0            0
EIC2          0            0
EIC3          0            0
EIC4          0            0
EIC5          0            0
EIC6          0            0
EIC7          0            0
EIC8          0            0
EIC9          0            0
EIC10     

>Use appropriate methods to clean the columns GEOCODE2, WEALTH1, ADI, DMA,and MSA.

I start exploring what are the values in these columns 

- **WEALTH1** means wealth rating. it should be numerical. So I think if there is an empty value is because there is no rating or raitn is 0
- **GEOCODE2** Geo Cluster Code indicating the level geography at which a record matches the census data. A nominal or symbolic field. Blank=No code has been assigned or did not match at any level.
- **ADI** meaning not known. not present in documentation
- **DMA** meaning not known. not present in documentation
- **MSA** meaning not known. not present in documentation

In [8]:
pd.DataFrame({"NaN": data[['GEOCODE2','WEALTH1','ADI','DMA','MSA']].isna().sum(), "EmptySpaces": data[['GEOCODE2','WEALTH1','ADI','DMA','MSA']].eq(' ').sum()}).sort_values('NaN', ascending = False) 

NaN  EmptySpaces
WEALTH1   44732            0
GEOCODE2    132          187
ADI         132            0
DMA         132            0
MSA         132            0

- **WEALTH1** a lot of nan but I think they should be 0. So no rating means 0
- **GEOCODE2** only 132 missing values. and 187 with an space
- **ADI** only 132 missing values 
- **DMA** only 132 missing values 
- **MSA** only 132 missing values 

In [9]:
data[['GEOCODE2']].value_counts()

GEOCODE2
A           34484
B           28505
D           16580
C           15524
              187
dtype: int64

In [11]:
data[['WEALTH1']].value_counts()

WEALTH1
9.0        7585
8.0        6793
7.0        6198
6.0        5825
5.0        5280
4.0        4810
3.0        4237
2.0        4085
1.0        3454
0.0        2413
dtype: int64

In [12]:
data[['ADI']].value_counts()

ADI  
13.0     7296
51.0     4622
65.0     3765
57.0     2836
105.0    2617
197.0    2075
109.0    2054
131.0    1978
67.0     1847
201.0    1772
107.0    1746
275.0    1717
233.0    1676
127.0    1651
329.0    1647
75.0     1588
241.0    1573
15.0     1472
83.0     1305
351.0    1193
111.0    1127
279.0    1075
213.0    1073
59.0     1000
157.0     996
129.0     955
181.0     924
281.0     863
0.0       843
367.0     813
271.0     807
577.0     773
263.0     745
335.0     696
63.0      686
209.0     669
179.0     656
455.0     645
291.0     634
337.0     613
269.0     611
215.0     602
245.0     597
315.0     585
71.0      582
221.0     576
307.0     573
203.0     558
133.0     554
277.0     548
77.0      542
319.0     506
383.0     501
301.0     488
427.0     472
53.0      422
199.0     413
17.0      413
113.0     412
187.0     411
303.0     409
321.0     406
177.0     393
243.0     392
451.0     392
361.0     388
211.0     385
173.0     366
175.0     365
185.0     358
61.0      356


In [13]:
data[['DMA']].value_counts()

DMA  
803.0    7296
602.0    4632
807.0    3765
505.0    2839
819.0    2588
539.0    2222
524.0    2079
623.0    2009
862.0    1847
753.0    1780
618.0    1772
613.0    1717
528.0    1651
534.0    1647
820.0    1638
609.0    1583
751.0    1566
825.0    1472
527.0    1316
560.0    1175
617.0    1127
517.0    1085
567.0    1073
563.0    1000
616.0     993
548.0     955
659.0     917
518.0     863
790.0     813
641.0     788
804.0     773
650.0     754
561.0     690
513.0     661
640.0     656
839.0     645
529.0     641
881.0     640
770.0     635
671.0     611
622.0     597
557.0     588
866.0     582
571.0     567
678.0     565
630.0     564
789.0     548
635.0     547
658.0     542
648.0     512
693.0     511
686.0     501
652.0     488
619.0     481
744.0     449
575.0     436
669.0     435
632.0     424
588.0     422
540.0     414
855.0     413
752.0     403
541.0     397
679.0     394
612.0     387
546.0     387
801.0     386
682.0     384
649.0     363
675.0     363
637.0     363


In [14]:
data[['MSA']].value_counts()

MSA   
0.0       21333
4480.0     4606
1600.0     4059
2160.0     2586
520.0      1685
6780.0     1525
5945.0     1500
7320.0     1472
7600.0     1457
3360.0     1414
6200.0     1409
8280.0     1378
7040.0     1360
5775.0     1255
1920.0     1217
5120.0     1195
6440.0     1076
7360.0     1046
2080.0      993
7400.0      929
6920.0      894
5000.0      824
3760.0      817
5960.0      791
3480.0      790
2680.0      771
5080.0      745
4120.0      722
1520.0      697
3120.0      691
8960.0      644
7240.0      619
6640.0      592
2800.0      584
3000.0      560
3600.0      543
5360.0      523
5880.0      507
640.0       487
5560.0      478
8520.0      474
3160.0      465
4920.0      456
4520.0      444
7510.0      427
8735.0      415
8200.0      395
7160.0      389
1000.0      387
3840.0      382
8560.0      363
2960.0      354
200.0       353
3320.0      341
2020.0      339
2840.0      334
7500.0      334
5920.0      333
440.0       311
8720.0      298
680.0       288
4900.0      286
2

#I create a new dataframe to keep the original one in case something is broken

In [15]:
data_2=data

In [16]:
#Clean WEALTH1
#Input 0 if NAN
data_2['WEALTH1'].fillna(0, inplace=True)

In [17]:
data[['WEALTH1']].value_counts()
#Checking that I have input the 0 as I wanted

WEALTH1
0.0        47145
9.0         7585
8.0         6793
7.0         6198
6.0         5825
5.0         5280
4.0         4810
3.0         4237
2.0         4085
1.0         3454
dtype: int64

In [18]:
#Clean GEOCODE2
#If missing or NAN then O from Other
data_2['GEOCODE2'].fillna('O', inplace=True)
data_2[['GEOCODE2']]= data_2[['GEOCODE2']].apply(lambda x: 'O' if x[0] == ' ' else x, axis=1)
data_2[['GEOCODE2']].value_counts()

GEOCODE2
A           34484
B           28505
D           16580
C           15524
O             319
dtype: int64

In [19]:
#ADI, DMA, MSA are codes so I can not replace Nan with means. All the values are whole. 
#I replace the values for -1 because is a code that is not existing at the moment in the distribution 
#Other option is to delete this rows in case the columns are important for the model. 
#132 rows are not many, if the nulls are all in the same rows. 

In [20]:
#as we can see in the next table all 132 rows has the same nans so is possible to delete this rows. 
data_2_1=data_2[['GEOCODE2','WEALTH1','ADI','DMA','MSA']]
data_2_1[data_2_1['ADI'].isnull()].head(132)

GEOCODE2  WEALTH1  ADI  DMA  MSA
577          O      0.0  NaN  NaN  NaN
1119         O      0.0  NaN  NaN  NaN
2250         O      3.0  NaN  NaN  NaN
3326         O      0.0  NaN  NaN  NaN
5558         O      8.0  NaN  NaN  NaN
5767         O      0.0  NaN  NaN  NaN
6615         O      0.0  NaN  NaN  NaN
7225         O      0.0  NaN  NaN  NaN
7883         O      0.0  NaN  NaN  NaN
9279         O      0.0  NaN  NaN  NaN
9795         O      0.0  NaN  NaN  NaN
11971        O      0.0  NaN  NaN  NaN
12205        O      0.0  NaN  NaN  NaN
12390        O      0.0  NaN  NaN  NaN
12572        O      9.0  NaN  NaN  NaN
12573        O      0.0  NaN  NaN  NaN
14117        O      0.0  NaN  NaN  NaN
17098        O      0.0  NaN  NaN  NaN
18650        O      0.0  NaN  NaN  NaN
18715        O      0.0  NaN  NaN  NaN
19179        O      0.0  NaN  NaN  NaN
19564        O      0.0  NaN  NaN  NaN
21019        O      0.0  NaN  NaN  NaN
21206        O      0.0  NaN  NaN  NaN
21266        O      0.0  NaN  NaN  NaN
22543        O      0.0  NaN  NaN  NaN
22846        O      0.0  NaN  NaN  NaN
23486        O      5.0  NaN  NaN  NaN
24765        O      2.0  NaN  NaN  NaN
26624        O      8.0  NaN  NaN  NaN
26801        O      0.0  NaN  NaN  NaN
26981        O      0.0  NaN  NaN  NaN
28012        O      0.0  NaN  NaN  NaN
28409        O      4.0  NaN  NaN  NaN
28771        O      0.0  NaN  NaN  NaN
29193        O      8.0  NaN  NaN  NaN
30066        O      0.0  NaN  NaN  NaN
30138        O      0.0  NaN  NaN  NaN
31184        O      8.0  NaN  NaN  NaN
31426        O      0.0  NaN  NaN  NaN
31507        O      0.0  NaN  NaN  NaN
31668        O      0.0  NaN  NaN  NaN
34233        O      8.0  NaN  NaN  NaN
34867        O      0.0  NaN  NaN  NaN
35664        O      0.0  NaN  NaN  NaN
35678        O      0.0  NaN  NaN  NaN
35787        O      0.0  NaN  NaN  NaN
36410        O      0.0  NaN  NaN  NaN
36990        O      0.0  NaN  NaN  NaN
37117        O      0.0  NaN  NaN  NaN
37324        O      2.0  NaN  NaN  NaN
37694        O      3.0  NaN  NaN  NaN
38561        O      3.0  NaN  NaN  NaN
40107        O      0.0  NaN  NaN  NaN
42014        O      0.0  NaN  NaN  NaN
42919        O      0.0  NaN  NaN  NaN
43696        O      0.0  NaN  NaN  NaN
44455        O      0.0  NaN  NaN  NaN
45030        O      6.0  NaN  NaN  NaN
47563        O      0.0  NaN  NaN  NaN
47585        O      0.0  NaN  NaN  NaN
48232        O      0.0  NaN  NaN  NaN
48347        O      9.0  NaN  NaN  NaN
48712        O      7.0  NaN  NaN  NaN
48759        O      9.0  NaN  NaN  NaN
49003        O      0.0  NaN  NaN  NaN
49843        O      0.0  NaN  NaN  NaN
50244        O      9.0  NaN  NaN  NaN
50322        O      0.0  NaN  NaN  NaN
50686        O      0.0  NaN  NaN  NaN
50918        O      0.0  NaN  NaN  NaN
51269        O      0.0  NaN  NaN  NaN
51420        O      8.0  NaN  NaN  NaN
51644        O      0.0  NaN  NaN  NaN
51872        O      0.0  NaN  NaN  NaN
53021        O      0.0  NaN  NaN  NaN
53398        O      0.0  NaN  NaN  NaN
54222        O      0.0  NaN  NaN  NaN
55138        O      0.0  NaN  NaN  NaN
56311        O      0.0  NaN  NaN  NaN
56799        O      0.0  NaN  NaN  NaN
57015        O      0.0  NaN  NaN  NaN
57481        O      0.0  NaN  NaN  NaN
59045        O      0.0  NaN  NaN  NaN
59698        O      0.0  NaN  NaN  NaN
60643        O      8.0  NaN  NaN  NaN
62209        O      4.0  NaN  NaN  NaN
64293        O      0.0  NaN  NaN  NaN
64356        O      0.0  NaN  NaN  NaN
65014        O      0.0  NaN  NaN  NaN
65256        O      0.0  NaN  NaN  NaN
65358        O      0.0  NaN  NaN  NaN
66553        O      0.0  NaN  NaN  NaN
66662        O      0.0  NaN  NaN  NaN
67211        O      0.0  NaN  NaN  NaN
68255        O      0.0  NaN  NaN  NaN
69377        O      0.0  NaN  NaN  NaN
69547        O      0.0  NaN  NaN  NaN
69943        O      0.0  NaN  NaN  NaN
70653        O      0.0  NaN  NaN  NaN
73569        O      0.0  NaN  NaN  NaN
75117        O      5.0  NaN

In [22]:
data_2['GENDER'] = data_2['GENDER'].fillna('F')

In [23]:
data_2[['GENDER']]= data_2[['GENDER']].apply(lambda x: 'other' if (x[0] != 'F' and x[0] != 'M') else x, axis=1)

In [24]:
data_2[['GENDER']].value_counts()

GENDER
F         51277
M         39094
other      5041
dtype: int64

In [25]:
drop_list = ['OSOURCE','ZIP']
drop_list 

['OSOURCE', 'ZIP']

In [26]:
#Drop the columns in drop_list and check the new number of columns in the datset
data_2=data.drop(columns=drop_list)
data_2.shape

(95412, 479)

# <span style="color:#7B241C">  03 EDA </span>

In [ ]:
#pip install ydata-profiling

In [ ]:
##Not working :(
from ydata_profiling import ProfileReport
ProfileReport(data_2)

In [28]:
pip install sweetviz

  Using cached sweetviz-2.1.4-py3-none-any.whl (15.1 MB)
  Using cached importlib_resources-5.12.0-py3-none-any.whl (36 kB)
Note: you may need to restart the kernel to use updated packages.


In [29]:
import sweetviz as sv

In [36]:

data['NOEXCH'] = data['NOEXCH'].astype(str) #This field creates problems
my_report = sv.analyze(data, pairwise_analysis='off') #Remove the pairwise analysis because there is too many variables
my_report.show_html() # Default arguments will generate to "SWEETVIZ_REPORT.html"

                                             |          | [  0%]   00:00 -> (? left)

Report SWEETVIZ_REPORT.html was generated! NOTEBOOK/COLAB USERS: the web browser MAY not pop up, regardless, the report IS saved in your notebook/colab files.


In [33]:
#pip install dtale

     --------------------------------------- 14.0/14.0 MB 25.2 MB/s eta 0:00:00
     --------------------------------------- 40.1/40.1 MB 13.3 MB/s eta 0:00:00
     ---------------------------------------- 65.9/65.9 MB 8.0 MB/s eta 0:00:00
     ---------------------------------------- 45.9/45.9 kB 2.2 MB/s eta 0:00:00
     ---------------------------------------- 219.7/219.7 kB ? eta 0:00:00
     ------------------------------------- 642.7/642.7 kB 19.8 MB/s eta 0:00:00
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
     ---------------------------------------- 7.2/7.2 MB 24.4 MB/s eta 0:00:00
     ---------------------------------------- 62.3/62.3 kB 3.3 MB/s eta 0:00:00
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
     --------------------------------------- 10.3/10.3 MB 25.1 MB/s eta 0:00:00
     ------------------------------------- 383.3/383.3 kB 23.3 MB/s eta 0:00:00
  

ERROR: Could not install packages due to an OSError: [WinError 5] Acceso denegado: 'C:\\Users\\luisj\\anaconda3\\Lib\\site-packages\\~cipy\\cluster\\_hierarchy.cp310-win_amd64.pyd'
Consider using the `--user` option or check the permissions.



In [34]:
dtale.show(data)

NameError: name 'dtale' is not defined

** First impresions from EDA **
Results in Sweetviz report

- OSOURCE has  896 unique values. Maybe it makes sense to delete this column
- Mailcode has two values B and space. only 1399 has B. Imbalance? or delete?
- PVastate 93954 values are empty
- NOEXCH 7 values are empty 
- RCINHSE again 93% values are empty
- Income are like categories not the amount values between 1-7

Need to split the dataset? is it too big to look single variables. First I would check what to predict. 